### PR TITLE
http-client-java, remove extra shadowing if all polymorphic models are in the same package

### PR DESCRIPTION
**Pull request by XiaofeiCao**
Remove the need to check `share-jsonserializable-code` setting for determining whether to shadow parent read-only(effectively) properties for JSON deserialization.

Should fix unnecessary read-only properties shadowing: https://apiview.dev/Assemblies/Review/2f43d81416e14779ac541fb65815d80c/c574461c3ae8458b981cdff776dde19a#com.azure.messaging.eventgrid.systemevents.AcsChatMessageDeletedEventData.public-String-getMessageId%28%29

### CODE REVIEW

File: packages/http-client-java/generator/http-client-generator-test/src/main/java/tsptest/enumnesteddiscriminator/models/Golden.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                } else if ("dna".equals(fieldName)) {
+                    dna = reader.getString();
+                } else if ("kind".equals(fieldName)) {
+                    kind = DogKind.fromString(reader.getString());
+                } else {
+                    reader.skipChildren();
+                }
+            }
+            Golden deserializedGolden = new Golden(weight);
+            deserializedGolden.setDna(dna);
```
Comment (XiaofeiCao) [inline packages/http-client-java/generator/http-client-generator-test/src/main/java/tsptest/enumnesteddiscriminator/models/Golden.java:88]: no shadowing for parent read-only properties, use package-private setter directly

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
You may want to check that the fluentgen test in autorest.java is still good.

**Comment by XiaofeiCao** [comment]
> You may want to check that the fluentgen test in autorest.java is still good.

Thanks, it did reveal a long-existing bug which happen to hide a bug introduced by this PR(bug + bug = no bug)...
Now everything works, at least our tests..

File: packages/http-client-java/generator/http-client-generator-core/src/main/java/com/microsoft/typespec/http/client/generator/core/util/ClientModelUtil.java [hunk 2]
```diff
@@ -867,12 +867,14 @@ public static ClientModel getDefiningModel(ClientModel model, ClientModelPropert
...
         // must be read-only and not appear in constructor
-        (property.isReadOnly() && !settings.isIncludeReadOnlyInConstructorArgs())
-            // immutable output model only has package-private setters, making its properties read-only
-            || isImmutableOutputModel(getDefiningModel(model, property), settings));
+        ((property.isReadOnly() && !settings.isIncludeReadOnlyInConstructorArgs())
+            // immutable output model only has package-private setters, making its properties effectively read-only
+            || (isImmutableOutputModel(getDefiningModel(model, property), settings))
+                // if property.isReadOnly(), whether it's required or not will not affect it being in constructor or not
+                // , thus only check when !property.isReadOnly() and the model is immutable output(effectively read-only)
+                && !(property.isRequired() && settings.isRequiredFieldsAsConstructorArgs()));
```
Comment (XiaofeiCao) [inline packages/http-client-java/generator/http-client-generator-core/src/main/java/com/microsoft/typespec/http/client/generator/core/util/ClientModelUtil.java:877]: a long existing bug not related to this PR but hides this PR's bug..

**Pull request merged**